src/main/java/com/minecolonies/coremod/client/render/worldevent/ColonyBlueprintRenderer.java

```diff
@@ -47,16 +56,34 @@ public class ColonyBlueprintRenderer
     private static final double CACHE_RESET_RANGE = 12.5F;
 
     /**
-     * The cached map of blueprints that are rendered.
+     * The cached list of blueprints to be rendered.
      */
-    private static Map<BlockPos, RenderData> blueprintCache = new HashMap<>();
+    private static Map<BlueprintCacheKey, List<BlockPos>> blueprintRenderCache = new HashMap<>();
+
+    /**
+     * The cached list of boxes to be rendered.
+     */
+    private static Map<BlockPos, BoxRenderData> boxRenderCache = new HashMap<>();
+
+    /**
+     * The cache of blueprint data.
+     */
+    private static final LoadingCache<BlueprintCacheKey, BlueprintPreviewData> blueprintDataCache = CacheBuilder.newBuilder()

```

Comment:
Probably join with `blueprintRenderCache` using tuple or sth like that, it's hard to understand when you have three caches
I see you want to cache loaded blueprints, but at least don't go over rendering cache expire duration (it makes no sense)

Reply:
There's three caches because they're caching different things.  Technically only this one is actually a cache, the others just store different kinds of "current data" that's recalculated infrequently.

The `blueprintDataCache` is the "real" cache, which keeps around a loaded `Blueprint` based on its path so that it doesn't need to keep reloading it from disk.  It's a `LoadingCache` so it gets populated implicitly by usage in the other cache.

The `blueprintRenderCache` is storing a list of blueprints that ought to be rendered right now, and where.  This is populated explicitly, so can't be a `LoadingCache`.

The data cache is expected to have a longer expiration than the render cache, because buildings go in and out of the render cache as the player moves around the world, but it will usually be the same ones coming back in that left a little while ago.

Reply:
render cache keeps it as long as you want to render it (regardless visibility, if you do visibility checks you should not cuz structurize do that itself)

and the box cache? doesnt each blueprint provide the info anyway? so why keep it standalone

Reply:
The structurize render cache still requires you to have a blueprint instance, or to give it a unique "slot" name (the latter being what the tools do).  This cache is what stores those instances to avoid having to come up with unique names (and caching those instead).  It also allows it to use the (perhaps more efficient) `renderStructureAtPosList` call instead of repeated `renderStructureAtPos`.

This doesn't do visibility checks for occlusion, but it does have an effective range beyond which it will not try to load and render blueprints, to avoid loading the entire colony.  (For context: this code is what does the "level 5 neighbour" rendering, as well as the builder's goggles.)

The box cache exists because it's pointless to recalculate that every render tick, so it does that up front at a slower pace instead.  As for why it's a separate cache, it's because each blueprint can generate multiple boxes, and some boxes may not have a blueprint to render, so they have different arity and different lookup keys.